[Linux][Runtime][IRGen] Mark metadata sections as retained and support section GC.

al45tair opened the pull request:
We want to support section GC, and the `__start` and `__stop` symbols don't count as roots for that (at least, not by default). Additionally, while GNU `ld` and `lld` will coalesce retained and non-retained sections, `gold` refuses to do so (see https://sourceware.org/bugzilla/show_bug.cgi?id=31415), which trips up the reflection code because that only expects to find a single section.

Turn on the `SHF_GNU_RETAIN` flag in `swiftrt.o`, and remove the existing workarounds for both of the above.

Also fix generation of module objects to be section GC compatible by marking the `__Swift_AST` symbol as used and finalising the module in IRGen.

Additionally, it turns out that we weren't initialising the `llvm::TargetOptions` properly and in particular the `UseInitArray` setting was wrong; this is masked by code in `gold` that converts `.ctors` and `.dtors` sections into `.init_array` and `.fini_array` respectively, but `lld` does not do this, and since Glibc no longer calls functions in these sections, that means that when linked with `lld`, any constructors or destructors generated by the Swift compiler don't run. Fix by setting `UseInitArray`.

rdar://123504095

al45tair committed:
[IRGen] Use llvm.used on ELF instead of llvm.compiler.used.

We changed to `llvm.compiler.used` because of the behaviour of `gold`,
which refuses to coalesce sections that have different `SHF_GNU_RETAIN`
flags, which causes problems with metadata.

Originally I thought we were going to have to generate two sections
with distinct names and have the runtime look for both of them, but
it turns out that the runtime only wants to see sections that have
`SHF_GNU_RETAIN` in any case.  It's really the reflection code that
is interested in being able to see non-retained sections.  The upshot
is that we don't need to use `llvm.compiler.used`; it's just fine if
we have duplicate sections, as long as the reflection code looks for
them when it's inspecting an ELF image.

This also means we no longer need to pass `-z nostart-stop-gc` to the
linker if we're using `lld`.

rdar://123504095

al45tair committed:
[Runtime] Tidy up section declarations slightly.

This also calls out the reflection sections, even though we don't
actually need to do anything special for them (as it turns out).

rdar://123504095

al45tair committed:
[RemoteInspection] Update ReflectionContext to cope with multiple sections.

When linking with `gold`, we can end up with two separate sections
for each of the reflection sections; this happens because `swiftrt.o`
declares them with `SHF_GNU_RETAIN` set (to stop them from vanishing
when section GC is enabled), but if the compiler is set to allow the
reflection information to be stripped then *it* will generate them
*without* `SHF_GNU_RETAIN`.

The other Linux linkers will coalesce the two sections and leave the
`SHF_GNU_RETAIN` flag set, but `gold` does not, and adds both of them
separately to the output.

The upshot is that we want `ReflectionContext` to look for both the
retained and non-retained sections and read the data from both of
them.

rdar://123504095

al45tair committed:
[IRGen] When creating a module object, mark `__Swift_AST` as used.

Without doing this, `__Swift_AST` gets stripped from the output.
We also need to call `IGM.finalize()` before `::performLLVM()` in the
`createSwiftModuleObjectFile()` function, so that we update the
section to mark it as retained.

rdar://123504095

al45tair committed:
[IRGen] Set `TargetOpts.UseInitArray` in `getIRTargetOptions()`.

IRGen sets up the `llvm::TargetOptions` by itself, rather than copying
them from the Clang instance (it has to do this, at present, because
Clang doesn't provide a way to get it to initialise one).

Unfortunately, it didn't set `UseInitArray`, which when linking with
GNU `ld` or `gold` is *fine* because those linkers automatically
convert `.ctors` and `.dtors` sections into `.init_array` and
`.fini_array` respectively.  *However*, `lld` does *not* do this
conversion, so when using `lld`, if the compiler generates a
constructor or destructor function, it won't be called(!)

The fix is to set `UseInitArray` properly; I chose to copy the setting
from Clang, just in case Clang knows of a reason why it shouldn't be
`true`.

rdar://123504095

al45tair committed:
[Test] Fix function_sections test to support different linkers.

`gold` and `lld` produce map files in different formats; this test
can and should work for both of them.

rdar://123504095

al45tair committed:
[Test] Add linker_overridden feature.

When testing different linkers, it's sometimes useful to run the tests
with `SWIFT_DRIVER_TEST_OPTIONS=" -use-ld=<linker>"`.  If we do this,
it will break a handful of tests because they expect the compiler driver
to choose an appropriate linker automatically.

To avoid having these fail, detect when someone has done this, and
set a new feature, `linker_overridden`, then mark the tests in question
with `UNSUPPORTED: linker_overridden`.

rdar://123504095

al45tair commented:
@swift-ci Please test

finagolfin commented:
Glad that you're trying to fix this properly, this closes #60406? Any plans to get this into 5.10 or will that have to rely on my old workaround of extra lld flags instead?

finagolfin commented on test/lit.cfg:
```diff
@@ -540,6 +540,11 @@ config.swift_driver_test_options += os.environ.get('SWIFT_DRIVER_TEST_OPTIONS',
+# Check if we overrode the linker; if we do, set a feature to say that, which
+# lets us disable tests that rely on the driver choosing the linker.
+if re.search(r'(?:^|[ \t])-use-ld=[^ \t]*', config.swift_driver_test_options):
+    config.available_features.add('linker_overridden')
```
This seems overly defensive, particularly since the tests handle multiple linkers. Any particular problem you're trying to avoid with linker overrides?

al45tair replied:
Not all of the tests handle overriding the linker, which is the point of this. Specifically, the tests I tagged with `linker_overridden` won't work with `-use-ld=<whatever>` in `SWIFT_DRIVER_TEST_OPTIONS` because they rely on the fact that the driver will choose an appropriate linker based on either the specified target or the LTO setting. If you force it with `-use-ld=<something>`, then it will try to use that thing instead of letting the driver pick, and those tests will then fail.

finagolfin replied:
Oh, I see: these are tests that already pass `-use-ld` for various reasons, so that expected linker cannot be overridden, got it.

kateinoigakukun commented:
@swift-ci test WebAssembly

al45tair commented:
> Glad that you're trying to fix this properly, this closes #60406?

Looks like it will, yes.

al45tair commented:
```
$ SWIFT_DRIVER_TEST_OPTIONS=" -Xlinker --gc-sections -use-ld=gold" utils/run-test --build-dir ../build/Ninja-RelWithDebInfoAssert/swift-linux-x86_64

...

Testing Time: 1005.77s
  Unsupported      : 2075
  Passed           : 8500
  Expectedly Failed:   33

2 warning(s) in tests
```

al45tair commented:
```
$ SWIFT_DRIVER_TEST_OPTIONS=" -Xlinker --gc-sections -use-ld=lld" utils/run-test --build-dir ../build/Ninja-RelWithDebInfoAssert/swift-linux-x86_64

...

Testing Time: 1005.83s
  Unsupported      : 2075
  Passed           : 8500
  Expectedly Failed:   33

2 warning(s) in tests
```

al45tair commented:
Hmmmm… not sure what this is all about:

```
Failed Tests (5):
  Swift(linux-x86_64) :: SourceKit/Compile/basic.swift
  Swift(linux-x86_64) :: SourceKit/CursorInfo/cursor_swiftinterface.swift
  Swift(linux-x86_64) :: SourceKit/DocSupport/doc_system_module_underscored.swift
  Swift(linux-x86_64) :: SourceKit/InterfaceGen/gen_swift_module.swift
  Swift(linux-x86_64) :: SourceKit/Macros/macro_option_set.swift
```

Will take a look at that tomorrow.

al45tair commented:
> Any plans to get this into 5.10 or will that have to rely on my old workaround of extra lld flags instead?

It won't be in 5.10.0. I think it'd be worth getting it into a point release though.

etcwilde reviewed (changes requested):
This seems like the right direction. I would like a bit more in-source documentation about what it's doing and why we need the two sets of sections, what it means for a symbol to be in one but not the other, etc...

al45tair committed:
[RemoteInspection] Use `bool` instead of `!!`.

Since this is adjacent to a `!=` operator, explicitly using `bool` seems clearer.

Co-authored-by: Evan Wilde <etceterawilde@gmail.com>

al45tair commented on include/swift/RemoteInspection/ReflectionContext.h:
```diff
@@ -601,7 +601,7 @@ class ReflectionContext
     bool Error = false;
     auto findELFSectionByName =
-        [&](llvm::StringRef Name) -> std::pair<RemoteRef<void>, uint64_t> {
+        [&](llvm::StringRef Name, bool Retained) -> std::pair<RemoteRef<void>, uint64_t> {
```
_Personally_ I think putting a default on a parameter of a lambda doesn't result in the most readable code ever — I think I'd rather write it out explicitly (which I'll have to do in the non-default case anyway).

As for matching, yes, the matching makes sense. The way this works is that there are a couple of compiler flags that enable the stripping of reflection metadata; those work by causing the compiler to _not_ mark the sections as used (i.e. not set the `SHF_GNU_RETAIN` flag). When the compiler builds an object file, it will create a set of reflection metadata sections, all with the same setting of the retain flag. In a normal build, I would expect to find _either_ all retained or all non-retained sections, _except_ where someone was using static libraries (which might be built with a different compiler flag setting). It would be unusual, I think, to vary the flag per object within a build.

If it helps clarify things slightly: the _runtime_ only cares about sections marked as retained, since those are the ones you've told the compiler you want to keep. That's why we only declare the retained ones in `swiftrt.o`. The code in RemoteInspection, by contrast, is interested in seeing what it can find out from the ELF binary, and in that case it _does_ make sense to look for both sets of sections.

al45tair replied:
I've added a large comment to ReflectionContext to explain all of this.

al45tair commented on include/swift/RemoteInspection/ReflectionContext.h:
```diff
@@ -649,48 +651,96 @@ class ReflectionContext
     SwiftObjectFileFormatELF ObjectFileFormat;
     auto FieldMdSec = findELFSectionByName(
-        ObjectFileFormat.getSectionName(ReflectionSectionKind::fieldmd));
+        ObjectFileFormat.getSectionName(ReflectionSectionKind::fieldmd), true);
```
Yeah, we only care about the reflection metadata sections here. I'm not hugely keen on making this change dependent on changes in LLVM; we could always do that separately later on if we think it's worthwhile.

al45tair committed:
[Test] Update section_asm to work on non-ELF platforms again.

On ELF platforms, the output will be slightly different to non-ELF
platforms.  The test should ideally run everywhere, however, so
we need to be able to distinguish these platforms by changing
the lit.cfg to add some extra variables.

rdar://123504095

al45tair committed:
[RemoteInspection] Add a big comment to ReflectionContext.h.

Add a comment to ReflectionContext.h explaining the situation regarding
`SHF_GNU_RETAIN` and the reflection metadata sections.

rdar://123504095

al45tair commented:
The SourceKit test failures are all of the form
```
Optimizer/Verifier.swift:22: Fatal error: instruction   %3 = struct $Int32 (%2 : $Builtin.Int32)        // user: %4
should conform to ForwardingInstruction
```
which seems to relate to something @eeckstein added on the 14th of February. They don't seem to be related to the changes in this PR.

al45tair commented:
@swift-ci test WebAssembly

etcwilde reviewed (approved):
LGTM. Thanks for the docs. There's so little actual documentation on a lot of the linkage behaviors that I think it's important to have it written down somewhere.

al45tair commented:
OK. Apparently the problem with the SourceKit tests is that the `sourcekitd-test` executable doesn't link `swiftrt.o`. That should be easy enough to fix.

al45tair commented:
@swift-ci Please test

al45tair commented:
@swift-ci test WebAssembly

aschwaighofer reviewed (commented):
The IRGen changes look good to me. I can't evaluate whether the runtime can handle having multiple sections of the same kind (e.g protocol conformances) but I guess we will find out.

etcwilde commented on tools/SourceKit/tools/sourcekitd-test/CMakeLists.txt:
```diff
@@ -18,9 +18,14 @@ else()
     dispatch
-    BlocksRuntime)
+    BlocksRuntime
+    "${arch_subdir}/swiftrt${CMAKE_C_OUTPUT_EXTENSION}"
```
how did this work before? 😨

al45tair replied:
That is an excellent question, and honestly, I'm not sure. It quite clearly needs `swiftrt.o` though, so…

al45tair replied:
Hmmm. Not sure this is the right fix, looking at it a bit more. There's an `AddSwiftSourceKit.cmake` file that I think _should_ already be adding these files, which is presumably how it worked before. Obviously that's going wrong somehow now.

al45tair replied:
Ohhhh… I bet it's linking the wrong `swiftrt.o`.

al45tair replied:
Trying to reproduce the problem locally. In the meantime I'm going to get rid of this change, and re-run CI so I can see the errors again.

etcwilde replied:
If we held CMake correctly, it would be using swiftc as the linker here instead of clang, which would then pick up swiftrt.o and the stdlib.

al45tair replied:
Indeed, instead of the complicated dance in `AddSwiftSourceKit.cmake` (which doesn't look entirely right to me; at the very least, I'm suspicious because it does nothing on Windows, which seems… wrong).

al45tair commented:
@swift-ci Please test

al45tair commented:
@swift-ci test WebAssembly

al45tair commented:
> I can't evaluate whether the runtime can handle having multiple sections of the same kind (e.g protocol conformances) but I guess we will find out.

I think _that_ was happening because `swiftrt.o` declared the sections _without_ `SHF_GNU_RETAIN` but `@llvm.used` changed to setting it (so we had different settings for it, so `gold` wasn't coalescing the sections any more). Turning on `SHF_GNU_RETAIN` in `swiftrt.o` didn't work previously because, among other things, the compiler _optionally_ sets it for reflection sections, and so we sometimes ended up with two sets of reflection sections and failed the reflection tests. The latter is fixed by one of the commits above by making the reflection code cope with having one or other, or both, sets of reflection sections. We don't have to do this in the runtime because the purpose of the flag that controls the compiler behaviour there is to enable stripping of reflection information, so `swiftrt.o` only needs to declare the retained version of the section.

So I don't _believe_ we should get multiple protocol conformance sections after these changes, since they _should_ all have `SHF_GNU_RETAIN` set.

al45tair commented:
@swift-ci Please test Windows platform

al45tair commented:
@swift-ci Please test Linux platform

al45tair commented:
(Rebased on main.)

al45tair committed:
[Runtime] Add the `retain` attribute to make the backtracer work.

Without `retain` here, we might remove the reference that pulls in
the backtracing support code.

rdar://123504095

al45tair commented:
I _think_ the SourceKit test failures _might_ be being caused by the `SwiftCompilerSources` being built with the host-installed compiler. If that happens, the protocol conformance sections will _not_ be marked as retained, so they could be discarded by the linker at which point the `self is ForwardingInstruction` test in `SwiftCompilerSources/Sources/Optimizer/Utilities/Verifier.swift` will fail.

al45tair committed:
[Test][SourceKit] Use the *builder*'s swiftrt.o when in hosttools.

We were linking with the newly built `swiftrt.o` when in hosttools mode,
which is wrong because the newly built `swiftrt.o` does not match the
compiler we were using for the `SwiftCompilerSources`.

This manifests as a failure in

  `SwiftCompilerSources/Sources/Optimizer/Utilities/Verifier.swift`

because `self is ForwardingInstruction` fails as we can't find the
protocol conformance records.

rdar://123504095

al45tair commented:
Awesome; I think that fixed the SourceKit failures. The remaining one isn't anything to do with this PR — that seems to be caused by #71993 (revert in #72149).

al45tair commented:
@swift-ci Please test

al45tair commented:
@swift-ci test WebAssembly

al45tair commented:
OK, so now there's some problem with the LLDB tests not finding the `__swiftast` section.

compnerd commented on include/swift/RemoteInspection/ReflectionContext.h:
```diff
@@ -649,48 +674,96 @@ class ReflectionContext
     if (Error)
       return false;
 
+    std::optional<uint32_t> result = false;
```
This is an odd initialisation. Why is the value optional when it is assigned a `0` to initialise it? This seems like it will always have a value.

al45tair replied:
That's a good point :-) I think I just blindly copied the return type of the function here without really thinking. And actually I'm now suspicious about the `return false` in other places in this function — I wonder if whoever wrote that meant to write `return {};`

compnerd commented on include/swift/RemoteInspection/ReflectionContext.h:
```diff
@@ -649,48 +674,96 @@ class ReflectionContext
+        TypeRefMdSec.first != nullptr ||
+        ReflStrMdSec.first != nullptr ||
+        ConformMdSec.first != nullptr ||
+        MPEnumMdSec.first != nullptr) {
```
Personally, I would rely on integer promotion and avoid the ` != nullptr` as it just adds noise.

al45tair replied:
:-) I was just keeping the existing style. (I would normally do exactly what you suggest.) Maybe we should change it.

compnerd commented on stdlib/public/runtime/SwiftRT-ELF-WASM.cpp:
```diff
@@ -28,40 +28,51 @@ static constexpr const void *__dso_handle = nullptr;
+# define DECLARE_EMPTY_METADATA_SECTION(name,attrs) __asm__("\t.section " #name ",\"" attrs "\"\n");
 #elif defined(__wasm__)
-# define DECLARE_EMPTY_METADATA_SECTION(name) __asm__("\t.section " #name ",\"R\",@\n");
+# define DECLARE_EMPTY_METADATA_SECTION(name,attrs) __asm__("\t.section " #name ",\"R\",@\n");
```
Nit: space after the `,` would be nice.

compnerd reviewed (approved):
I'm not sure how I feel about the separated sections, but this seems better than the reliance on `-z nostart-stop-gc`.

al45tair commented:
> I'm not sure how I feel about the separated sections

We'll only end up with two sets of sections if someone links together code that has different settings for the retain-reflection-metadata compiler flags, _and_ they're using `gold` to do their link. If the link with `lld` (or in theory GNU `ld`, though we don't support that), that won't happen, and if they use consistent settings it won't happen either.

al45tair committed:
[Runtime] Tidy up a couple of minor nits.

This doesn't change any code, just makes things look slightly
neater.

rdar://123504095

al45tair committed:
[RemoteInspection] Change "return false" to "return {}".

There are a number of instances in `ReflectionContext.h` where we
are doing `return false` with an `std::optional<...>` where it seems
we really mean to return an empty optional instead.

(The way to do this is either `return {}` or `return std::nullopt`.)

rdar://123504095

al45tair commented:
Finally managed to reproduce the LLDB problem (and work out why I couldn't reproduce the SourceKit problem locally). I'd upgraded `gold` on my dev box. Downgrading it reproduces the problem. Doesn't make it a `gold` problem, mind — we still need to fix it.

al45tair commented:
LLDB problem is caused by a bug in LLDB (rdar://124467787) that triggers if it tries to look up symbols in a binary that has multiple `.text` sections. That only happens with `ld.gold` from binutils-2.35 or earlier; newer `ld.gold` will merge the sections in question together and the problem goes away.

al45tair committed:
[Build] Detect ld.gold version and prefer lld if gold is too old.

If we're on a system that has ld.gold 2.35 or earlier, we want to use
lld instead because otherwise we end up with duplicate sections in the
output.

rdar://123504095

al45tair commented:
@swift-ci Please test

etcwilde reviewed (commented):
I wonder if we should push this up into the `apple/llvm-project` repo and then default `SWIFT_USE_LINKER_default` to some combination of `LLVM_USE_LINKER`/`LLVM_ENABLE_LLD` (probably just `LLVM_USE_LINKER` is fine since `LLVM_ENABLE_LLD` implies `LLVM_USE_LINKER` is `lld`.)

We should also be able to remove the Amazon Linux distro name check, which would be good too.

al45tair commented:
Hmmm. That fixed the `.text` problem that was breaking LLDB, but it looks like there's some more LLDB breakage besides.

al45tair commented:
(For those following this PR, I'm still working on the LLDB issues. One of them is fixed here: https://github.com/llvm/llvm-project/pull/90099, but there are some more things going wrong for some reason.)

al45tair commented:
OK, so the "more things going wrong" are actually _the same problem_ as the one fixed in llvm/llvm-project#90099. a014bd2 only works around it for `libswiftCore.dylib` (and other things built by the Swift build system), but it _doesn't_ work around it for LLDB's tests.

What was happening there, FWIW, was that a number of tests set a breakpoint on `main`, which, if there are two `.text` sections and LLDB doesn't have the fix from llvm/llvm-project#90099, typically results in a breakpoint on `main` (as expected) but also one on `__start` (because _that_ symbol is at the beginning of the _other_ `.text` section). Then when you run the program, it stops at `__start` instead of stopping in `main`, and since `__start` isn't a Swift frame, the language defaults to C, so evaluating Swift expressions doesn't work (they don't parse as C), and it's also not the expected frame, so the variables the tests expect to be able to inspect aren't there.

al45tair commented:
apple/llvm-project#8661

@swift-ci Please test

drodriguez commented on test/LinkerSections/function_sections.swift:
```diff
@@ -1,10 +1,13 @@
 // RUN: %target-build-swift -Xfrontend -function-sections -emit-module -emit-library -static -parse-stdlib %S/Inputs/FunctionSections.swift
-// RUN: %target-build-swift -Xlinker --gc-sections -Xlinker -Map=%t/../../FunctionSections.map -I%t/../.. -L%t/../.. -lFunctionSections %S/Inputs/FunctionSectionsUse.swift
-// RUN: %FileCheck %s < %t/../../FunctionSections.map
+// RUN: %target-build-swift -Xlinker -v -Xlinker --gc-sections -Xlinker -Map=%t/../../FunctionSections.map -I%t/../.. -L%t/../.. -lFunctionSections %S/Inputs/FunctionSectionsUse.swift 2>&1 | sed 's/.*\(gold\|LLD\).*/\1/g' | tr "[:lower:]" "[:upper:]" > %t/../../Linker.txt
```
This change takes the stdout and stderr of `%target-build-stdlib` pass __each line__ through `sed` that replaces lines that say either `gold` or `LLD` with just those word, and convert everything to upper case. Then the result of all that is passed as `--check-prefix`?

What happens when `-v` prints more than one line? All the lines will be printed and passed to as extra arguments to `%FileCheck`?

```
$ echo -e 'foo\nbar\nthis will say LLD\nbaz' | sed 's/.*\(gold\|LLD\).*/\1/g'                                                                                                                      foo
bar
LLD
baz
```

Am I missing something?

drodriguez replied:
#73470 73470
